### PR TITLE
planner: cherry-pick enforce-mpp to v5.0

### PR DESCRIPTION
**Pull request by LittleFall**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

5.0 needs enforce-mpp.

### What is changed and how it works?

cherry-pick these prs to 5.0:
1. https://github.com/pingcap/tidb/pull/24157 CopTiFlashConcurrencyFactor 
2. https://github.com/pingcap/tidb/pull/24516 support allow_mpp=2
3. https://github.com/pingcap/tidb/pull/25302 revert 24516 and support tidb_enforce_mpp and add warning
4. https://github.com/pingcap/tidb/pull/25369 add config
5. https://github.com/pingcap/tidb/pull/25450 enforce projection push down
6. https://github.com/pingcap/tidb/pull/25553 fix warning
7. https://github.com/pingcap/tidb/pull/26246 add warning of join

### Check List 

Tests 

- [x] Unit test
- [x] Integration test

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

- Support set `tidb_enforce_mpp=1` to enforce use mpp mode.

### CODE REVIEW

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -3532,22 +3532,3 @@ func (s *testIntegrationSerialSuite) TestMergeContinuousSelections(c *C) {
 		res.Check(testkit.Rows(output[i].Plan...))
 	}
 }
-
-func (s *testIntegrationSuite) TestIssue23839(c *C) {
```
Comment (fzhedu) [inline planner/core/integration_test.go:3536]: why delete this?

**Review by hanfei1991 — APPROVED** [review]

File: planner/core/integration_test.go (hunk 1 above)
Comment (LittleFall) [reply planner/core/integration_test.go]: my fault, reverted.

**Review by fzhedu — APPROVED** [review]
/LGTM

**Pull request merged**